File: mediaplayer_test/app/src/main/java/com/daeseong/mediaplayer_test/Mp3Player.java
package com.daeseong.mediaplayer_test;

import android.content.Context;
import android.media.MediaPlayer;
import android.util.Log;

public class Mp3Player {

    private static final String TAG = Mp3Player.class.getName();

    private static Mp3Player instance;
    private MediaPlayer mediaPlayer;
    private OnMediaPlayerListener onMediaPlayerListener;

    private Mp3Player() {
        mediaPlayer = new MediaPlayer();
    }

    public static Mp3Player getInstance() {
        if (instance == null) {
            instance = new Mp3Player();
        }
        return instance;
    }

    public void release() {
        try {
            removeListener();

            if (mediaPlayer != null) {
                mediaPlayer.stop();
                mediaPlayer.reset();
                mediaPlayer.release();
                mediaPlayer = null;
            }
        } catch (Exception ex) {
            Log.e(TAG, ex.getMessage());
        }
    }

    public void play(String sPath, final OnMediaPlayerListener onMediaPlayerListener) {
        try {
            this.onMediaPlayerListener = onMediaPlayerListener;

            mediaPlayer.reset();
            mediaPlayer.setDataSource(sPath);
            mediaPlayer.prepare();

            setMediaPlayerListeners();
        } catch (Exception ex) {
            handleMediaPlayerError(ex);
        }
    }

    public void play(Context context, final OnMediaPlayerListener onMediaPlayerListener) {
        try {
            this.onMediaPlayerListener = onMediaPlayerListener;

            mediaPlayer = MediaPlayer.create(context, R.raw.a);

            setMediaPlayerListeners();
        } catch (Exception ex) {
            handleMediaPlayerError(ex);
        }
    }

    private void setMediaPlayerListeners() {
        mediaPlayer.setOnCompletionListener(new MediaPlayer.OnCompletionListener() {
            @Override
            public void onCompletion(MediaPlayer mp) {
                if (onMediaPlayerListener != null) {
                    onMediaPlayerListener.onCompletion(true);
                }
            }
        });

        mediaPlayer.setOnPreparedListener(new MediaPlayer.OnPreparedListener() {
            @Override
            public void onPrepared(MediaPlayer mp) {
                if (onMediaPlayerListener != null) {
                    onMediaPlayerListener.onPrepared(mp.getDuration());
                }
            }
        });

        mediaPlayer.setOnErrorListener(new MediaPlayer.OnErrorListener() {
            @Override
            public boolean onError(MediaPlayer mp, int what, int extra) {
                handleMediaPlayerError(new Exception("MediaPlayer error"));
                return false;
            }
        });
    }

    public void start() {
        if (mediaPlayer != null) {
            mediaPlayer.start();
        }
    }

    public void stop() {
        if (mediaPlayer != null && mediaPlayer.isPlaying()) {
            mediaPlayer.stop();
            mediaPlayer.reset();
        }
    }

    public void pause() {
        if (mediaPlayer != null && mediaPlayer.isPlaying()) {
            mediaPlayer.pause();
        }
    }

    public int getCurrentPosition() {
        if (mediaPlayer != null) return mediaPlayer.getCurrentPosition();
        else return 0;
    }

    public int getDuration() {
        if (mediaPlayer != null) return mediaPlayer.getDuration();
        else return 0;
    }

    public boolean isPlaying() {
        return mediaPlayer != null && mediaPlayer.isPlaying();
    }

    public void seekTo(int position) {
        if (mediaPlayer != null) {
            mediaPlayer.seekTo(position);
        }
    }

    private void removeListener() {
        onMediaPlayerListener = null;
    }

    private void handleMediaPlayerError(Exception ex) {
        if (onMediaPlayerListener != null) {
            onMediaPlayerListener.onCompletion(false);
        }
        Log.e(TAG, ex.getMessage());
    }

    public interface OnMediaPlayerListener {
        void onCompletion(boolean bComplete);
        void onPrepared(int mDuration);
    }
}

File: mediaplayer_test(kotlin)/app/src/main/java/com/daeseong/mediaplayer_test/Mp3Player.kt
package com.daeseong.mediaplayer_test

import android.content.Context
import android.media.MediaPlayer
import android.util.Log

class Mp3Player private constructor() {

    private val tag = Mp3Player::class.java.name

    private var mediaPlayer: MediaPlayer? = null
    private var onMediaPlayerListener: OnMediaPlayerListener? = null

    fun release() {
        try {
            removeListener()

            mediaPlayer?.let {
                it.stop()
                it.reset()
                it.release()
            }
            mediaPlayer = null
        } catch (ex: Exception) {
            Log.e(tag, ex.message ?: "")
        }
    }

    fun play(sPath: String, onMediaPlayerListener: OnMediaPlayerListener) {
        try {
            this.onMediaPlayerListener = onMediaPlayerListener

            mediaPlayer?.reset()
            mediaPlayer?.setDataSource(sPath)
            mediaPlayer?.prepare()

            setMediaPlayerListeners()
        } catch (ex: Exception) {
            handleMediaPlayerError(ex)
        }
    }

    fun play(context: Context, onMediaPlayerListener: OnMediaPlayerListener) {
        try {
            this.onMediaPlayerListener = onMediaPlayerListener

            mediaPlayer = MediaPlayer.create(context, R.raw.a)

            setMediaPlayerListeners()
        } catch (ex: Exception) {
            handleMediaPlayerError(ex)
        }
    }

    private fun setMediaPlayerListeners() {
        mediaPlayer?.setOnCompletionListener {
            onMediaPlayerListener?.onCompletion(true)
        }

        mediaPlayer?.setOnPreparedListener {
            onMediaPlayerListener?.onPrepared(it.duration)
        }

        mediaPlayer?.setOnErrorListener { _, _, _ ->
            handleMediaPlayerError(Exception("MediaPlayer error"))
            false
        }
    }

    fun start() {
        mediaPlayer?.start()
    }

    fun stop() {
        if (mediaPlayer?.isPlaying == true) {
            mediaPlayer?.stop()
            mediaPlayer?.reset()
        }
    }

    fun pause() {
        if (mediaPlayer?.isPlaying == true) {
            mediaPlayer?.pause()
        }
    }

    fun getCurrentPosition(): Int {
        return mediaPlayer?.currentPosition ?: 0
    }

    fun getDuration(): Int {
        return mediaPlayer?.duration ?: 0
    }

    fun isPlaying(): Boolean {
        return mediaPlayer?.isPlaying == true
    }

    fun seekTo(position: Int) {
        mediaPlayer?.seekTo(position)
    }

    private fun removeListener() {
        onMediaPlayerListener = null
    }

    private fun handleMediaPlayerError(ex: Exception) {
        onMediaPlayerListener?.onCompletion(false)
        Log.e(tag, ex.message ?: "")
    }

    interface OnMediaPlayerListener {
        fun onCompletion(bComplete: Boolean)
        fun onPrepared(mDuration: Int)
    }

    companion object {
        private var instance: Mp3Player? = null

        @JvmStatic
        fun getInstance(): Mp3Player {
            if (instance == null) {
                instance = Mp3Player()
            }
            return instance!!
        }
    }
}